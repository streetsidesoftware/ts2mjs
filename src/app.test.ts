import { Command, CommanderError } from 'commander';
import { promises as fs } from 'fs';
import { relative, join } from 'path';
import { format } from 'util';
import { afterEach, describe, expect, type Mock, test, vi } from 'vitest';

import { run as runApp } from './app.js';
import { resolveTempUnique } from './test.util.js';

// const oc = (a: object) => expect.objectContaining(a);
const sc = (s: string) => expect.stringContaining(s);
const sm = (s: string | RegExp) => expect.stringMatching(s);

describe('app', () => {
    afterEach(() => {
        vi.clearAllMocks();
    });

    test.each`
        args                                                            | expectedOutputs
        ${['.', '--root=fixtures/sample/lib', '--output=temp']}         | ${[sm(/index.js\b.*\bindex.mjs Generated/), sc('index.js.map - copy'), sc('done.')]}
        ${['.', '--cwd=fixtures/sample/lib', '--output=../../../temp']} | ${[sm(/index.js\b.*\bindex.mjs Generated/), sc('index.js.map - copy'), sc('done.')]}
        ${['fixtures/sample/lib']}                                      | ${[sm(/index.js\b.*\bindex.mjs Generated/), sc('done.')]}
        ${['fixtures/sample/lib']}                                      | ${[sm(/index.js\b.*\bindex.mjs Generated/), sc('done.')]}
        ${['not_found', '--no-must-find-files']}                        | ${[sc('done.')]}
    `('run (--dry-run) $args', async ({ args, expectedOutputs }: { args: string[]; expectedOutputs: unknown[] }) => {
        const argv = genArgv(args);
        const context = createRunContext();
        await expect(run(argv, context)).resolves.toBeUndefined();
        const output = context.output({ sort: true });
        for (const expected of expectedOutputs) {
            expect(output).toEqual(expected);
        }
        expect(output).toMatchSnapshot();
    });

    test.each`
        args                                                          | expectedOutputs
        ${['.', '--root=fixtures/sample', '--exclude=**/lib-cjs/**']} | ${[sm(/index.js\b.*\bindex.mjs Generated/), sc('index.js.map - copy'), sc('done.')]}
    `('run (actual) $args', async ({ args, expectedOutputs }: { args: string[]; expectedOutputs: unknown[] }) => {
        const tempDir = relative(process.cwd(), resolveTempUnique());
        const argv = genArgv([...args, `--output=${tempDir}`], { dryRun: false });
        const context = createRunContext();
        await expect(run(argv, context)).resolves.toBeUndefined();
        const output = context.output({ sort: true, replacements: [tempDir, 'temp'] });
        for (const expected of expectedOutputs) {
            expect(output).toEqual(expected);
        }
        expect(output).toMatchSnapshot();
    });

    test.each`
        args                                                                          | expected
        ${['.', '--root=fixtures/sample/lib/database', '--cjs']}                      | ${sc('[error]: Error: Import of a file outside of the root.')}
        ${['.', '--root=fixtures/sample/lib/database', '--cjs', '--no-enforce-root']} | ${sc('[error]: Warning: Import of a file outside of the root. Import: (../types.js) Source: (fetch.d.ts)')}
    `('run (actual) (errors and warnings) $args', async ({ args, expected }) => {
        const tempDir = relative(process.cwd(), resolveTempUnique());
        const argv = genArgv([...args, `--output=${tempDir}`], { dryRun: false });
        const context = createRunContext();
        await expect(run(argv, context)).resolves.toBeUndefined();
        const output = context.output({ sort: true, replacements: [tempDir, 'temp'] });
        expect(output).toEqual(expected);
    });

    test('run (actual) --remove-source', async () => {
        // Create temporary test files that can be safely deleted
        const tempSourceDir = resolveTempUnique();
        const tempOutputDir = resolveTempUnique();

        // Create test source files
        await fs.mkdir(tempSourceDir, { recursive: true });
        await fs.writeFile(join(tempSourceDir, 'test.js'), 'export const test = "hello";');
        await fs.writeFile(join(tempSourceDir, 'main.js'), 'import { test } from "./test.js";');

        // Run conversion with --remove-source
        const argv = genArgv(['.', `--root=${tempSourceDir}`, `--output=${tempOutputDir}`, '--remove-source'], {
            dryRun: false,
        });
        const context = createRunContext();
        await expect(run(argv, context)).resolves.toBeUndefined();

        const output = context.output({ sort: true });

        // Verify outputs
        expect(output).toEqual(sc('Generated'));
        expect(output).toEqual(sc('removed'));
        expect(output).toEqual(sc('done.'));

        // Verify original files were removed
        await expect(fs.access(join(tempSourceDir, 'test.js'))).rejects.toThrow();
        await expect(fs.access(join(tempSourceDir, 'main.js'))).rejects.toThrow();

        // Verify new files were created
        await expect(fs.access(join(tempOutputDir, 'test.mjs'))).resolves.toBeUndefined();
        await expect(fs.access(join(tempOutputDir, 'main.mjs'))).resolves.toBeUndefined();

        // Verify imports were updated
        const mainContent = await fs.readFile(join(tempOutputDir, 'main.mjs'), 'utf8');
        expect(mainContent).toEqual(sc('./test.mjs'));
    });

    test.each`
        args
        ${'--help'}
        ${'_not_found_'}
    `('run error $args', async ({ args }) => {
        const argv = genArgv(args);
        const context = createRunContext();
        await expect(run(argv, context)).rejects.toBeInstanceOf(CommanderError);
        expect(context.output()).toMatchSnapshot();
    });
});

function genArgv(args: string | string[], { dryRun = true, color = false, verbose = true } = {}): string[] {
    args = typeof args === 'string' ? [args] : args;
    const defaultArgs: string[] = [];
    dryRun && defaultArgs.push('--dry-run');
    defaultArgs.push(color ? '--color' : '--no-color');
    verbose && defaultArgs.push('--verbose');
    const argv: string[] = [process.argv[0], 'bin.mjs', ...defaultArgs, ...args];
    return argv;
}

interface FormatOptions {
    sort?: boolean;
    replacements?: [oldVal: string, newVal: string];
}

function formatOutput(logger: AppLogger, { sort, replacements }: FormatOptions = {}): string {
    return `\
Console Output:
${prefix(outputToString(logger.log.mock.calls, sort, replacements), '[log]: ')}
${prefix(outputToString(logger.warn.mock.calls, sort, replacements), '[warn]: ')}
${prefix(outputToString(logger.error.mock.calls, sort, replacements), '[error]: ')}
${prefix(outputToString(logger.stdout.mock.calls, false, undefined), '[stdout]: ')}
${prefix(outputToString(logger.stderr.mock.calls, false, undefined), '[stderr]: ')}
`;
}

function prefix(text: string, pfx: string): string {
    return pfx + text.split('\n').join('\n' + pfx);
}

function outputToString(calls: unknown[][], sort = false, replacements?: [oldVal: string, newVal: string]): string {
    const callLines = calls.map((call) => call.join('|'));
    if (sort) {
        callLines.sort();
    }
    const output = callLines.join('\n');
    const adjusted = replacements ? replaceAll(output, replacements[0], replacements[1]) : output;
    return adjusted.replace(/\\/g, '/');
}

function replaceAll(text: string, replace: string, withValue: string): string {
    const result = text.split(replace).join(withValue);
    if (/\.\d{3}Z/.test(result)) {
        process.stderr.write(format('%o', { text, replace, withValue, result }));
    }
    return result;
}

type LogFn = (message: string) => void;

interface AppLogger {
    log: Mock<LogFn>;
    warn: Mock<LogFn>;
    error: Mock<LogFn>;
    stderr: Mock<LogFn>;
    stdout: Mock<LogFn>;
}

function getAppLogger(): AppLogger {
    return {
        log: vi.fn<LogFn>(),
        warn: vi.fn<LogFn>(),
        error: vi.fn<LogFn>(),
        stderr: vi.fn<LogFn>(),
        stdout: vi.fn<LogFn>(),
    };
}

function run(argv: string[], context: RunContext) {
    return runApp(argv, context.program, context.logger);
}

interface RunContext {
    program: Command;
    logger: AppLogger;
    output: (options?: FormatOptions) => string;
}

function createRunContext(): RunContext {
    const logger = getAppLogger();
    const program = new Command();
    program.configureOutput({
        writeOut: logger.stdout,
        writeErr: logger.stderr,
    });
    program.exitOverride((e) => {
        throw e;
    });

    function output(opts?: FormatOptions) {
        return formatOutput(logger, opts);
    }

    return {
        logger,
        program,
        output,
    };
}
