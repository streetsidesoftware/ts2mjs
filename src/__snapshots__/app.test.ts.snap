// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`app > run (--dry-run) [ '.', …(2) ] 1`] = `
"Console Output:
[log]: done.
[log]: fixtures/sample/lib/app.d.ts -> temp/app.d.mts Generated
[log]: fixtures/sample/lib/app.d.ts.map - copy
[log]: fixtures/sample/lib/app.d.ts.map -> temp/app.d.mts.map Generated
[log]: fixtures/sample/lib/app.js -> temp/app.mjs Generated
[log]: fixtures/sample/lib/app.js.map - copy
[log]: fixtures/sample/lib/app.js.map -> temp/app.mjs.map Generated
[log]: fixtures/sample/lib/constants.d.ts -> temp/constants.d.mts Generated
[log]: fixtures/sample/lib/constants.d.ts.map - copy
[log]: fixtures/sample/lib/constants.d.ts.map -> temp/constants.d.mts.map Generated
[log]: fixtures/sample/lib/constants.js -> temp/constants.mjs Generated
[log]: fixtures/sample/lib/constants.js.map - copy
[log]: fixtures/sample/lib/constants.js.map -> temp/constants.mjs.map Generated
[log]: fixtures/sample/lib/database/fetch.d.ts -> temp/database/fetch.d.mts Generated
[log]: fixtures/sample/lib/database/fetch.d.ts.map - copy
[log]: fixtures/sample/lib/database/fetch.d.ts.map -> temp/database/fetch.d.mts.map Generated
[log]: fixtures/sample/lib/database/fetch.js -> temp/database/fetch.mjs Generated
[log]: fixtures/sample/lib/database/fetch.js.map - copy
[log]: fixtures/sample/lib/database/fetch.js.map -> temp/database/fetch.mjs.map Generated
[log]: fixtures/sample/lib/database/index.d.ts -> temp/database/index.d.mts Generated
[log]: fixtures/sample/lib/database/index.d.ts.map - copy
[log]: fixtures/sample/lib/database/index.d.ts.map -> temp/database/index.d.mts.map Generated
[log]: fixtures/sample/lib/database/index.js -> temp/database/index.mjs Generated
[log]: fixtures/sample/lib/database/index.js.map - copy
[log]: fixtures/sample/lib/database/index.js.map -> temp/database/index.mjs.map Generated
[log]: fixtures/sample/lib/index.d.ts -> temp/index.d.mts Generated
[log]: fixtures/sample/lib/index.d.ts.map - copy
[log]: fixtures/sample/lib/index.d.ts.map -> temp/index.d.mts.map Generated
[log]: fixtures/sample/lib/index.js -> temp/index.mjs Generated
[log]: fixtures/sample/lib/index.js.map - copy
[log]: fixtures/sample/lib/index.js.map -> temp/index.mjs.map Generated
[log]: fixtures/sample/lib/lookup.d.ts -> temp/lookup.d.mts Generated
[log]: fixtures/sample/lib/lookup.d.ts.map - copy
[log]: fixtures/sample/lib/lookup.d.ts.map -> temp/lookup.d.mts.map Generated
[log]: fixtures/sample/lib/lookup.js -> temp/lookup.mjs Generated
[log]: fixtures/sample/lib/lookup.js.map - copy
[log]: fixtures/sample/lib/lookup.js.map -> temp/lookup.mjs.map Generated
[log]: fixtures/sample/lib/types.d.ts -> temp/types.d.mts Generated
[log]: fixtures/sample/lib/types.d.ts.map - copy
[log]: fixtures/sample/lib/types.d.ts.map -> temp/types.d.mts.map Generated
[log]: fixtures/sample/lib/types.js -> temp/types.mjs Generated
[log]: fixtures/sample/lib/types.js.map - copy
[log]: fixtures/sample/lib/types.js.map -> temp/types.mjs.map Generated
[warn]: 
[error]: 
[stdout]: 
[stderr]: 
"
`;

exports[`app > run (--dry-run) [ '.', …(2) ] 2`] = `
"Console Output:
[log]: app.d.ts -> ../../../temp/fixtures/sample/lib/app.d.mts Generated
[log]: app.d.ts.map - copy
[log]: app.d.ts.map -> ../../../temp/fixtures/sample/lib/app.d.mts.map Generated
[log]: app.js -> ../../../temp/fixtures/sample/lib/app.mjs Generated
[log]: app.js.map - copy
[log]: app.js.map -> ../../../temp/fixtures/sample/lib/app.mjs.map Generated
[log]: constants.d.ts -> ../../../temp/fixtures/sample/lib/constants.d.mts Generated
[log]: constants.d.ts.map - copy
[log]: constants.d.ts.map -> ../../../temp/fixtures/sample/lib/constants.d.mts.map Generated
[log]: constants.js -> ../../../temp/fixtures/sample/lib/constants.mjs Generated
[log]: constants.js.map - copy
[log]: constants.js.map -> ../../../temp/fixtures/sample/lib/constants.mjs.map Generated
[log]: database/fetch.d.ts -> ../../../temp/fixtures/sample/lib/database/fetch.d.mts Generated
[log]: database/fetch.d.ts.map - copy
[log]: database/fetch.d.ts.map -> ../../../temp/fixtures/sample/lib/database/fetch.d.mts.map Generated
[log]: database/fetch.js -> ../../../temp/fixtures/sample/lib/database/fetch.mjs Generated
[log]: database/fetch.js.map - copy
[log]: database/fetch.js.map -> ../../../temp/fixtures/sample/lib/database/fetch.mjs.map Generated
[log]: database/index.d.ts -> ../../../temp/fixtures/sample/lib/database/index.d.mts Generated
[log]: database/index.d.ts.map - copy
[log]: database/index.d.ts.map -> ../../../temp/fixtures/sample/lib/database/index.d.mts.map Generated
[log]: database/index.js -> ../../../temp/fixtures/sample/lib/database/index.mjs Generated
[log]: database/index.js.map - copy
[log]: database/index.js.map -> ../../../temp/fixtures/sample/lib/database/index.mjs.map Generated
[log]: done.
[log]: index.d.ts -> ../../../temp/fixtures/sample/lib/index.d.mts Generated
[log]: index.d.ts.map - copy
[log]: index.d.ts.map -> ../../../temp/fixtures/sample/lib/index.d.mts.map Generated
[log]: index.js -> ../../../temp/fixtures/sample/lib/index.mjs Generated
[log]: index.js.map - copy
[log]: index.js.map -> ../../../temp/fixtures/sample/lib/index.mjs.map Generated
[log]: lookup.d.ts -> ../../../temp/fixtures/sample/lib/lookup.d.mts Generated
[log]: lookup.d.ts.map - copy
[log]: lookup.d.ts.map -> ../../../temp/fixtures/sample/lib/lookup.d.mts.map Generated
[log]: lookup.js -> ../../../temp/fixtures/sample/lib/lookup.mjs Generated
[log]: lookup.js.map - copy
[log]: lookup.js.map -> ../../../temp/fixtures/sample/lib/lookup.mjs.map Generated
[log]: types.d.ts -> ../../../temp/fixtures/sample/lib/types.d.mts Generated
[log]: types.d.ts.map - copy
[log]: types.d.ts.map -> ../../../temp/fixtures/sample/lib/types.d.mts.map Generated
[log]: types.js -> ../../../temp/fixtures/sample/lib/types.mjs Generated
[log]: types.js.map - copy
[log]: types.js.map -> ../../../temp/fixtures/sample/lib/types.mjs.map Generated
[warn]: 
[error]: 
[stdout]: 
[stderr]: 
"
`;

exports[`app > run (--dry-run) [ 'fixtures/sample/lib' ] 1`] = `
"Console Output:
[log]: done.
[log]: fixtures/sample/lib/app.d.ts -> fixtures/sample/lib/app.d.mts Generated
[log]: fixtures/sample/lib/app.d.ts.map -> fixtures/sample/lib/app.d.mts.map Generated
[log]: fixtures/sample/lib/app.js -> fixtures/sample/lib/app.mjs Generated
[log]: fixtures/sample/lib/app.js.map -> fixtures/sample/lib/app.mjs.map Generated
[log]: fixtures/sample/lib/constants.d.ts -> fixtures/sample/lib/constants.d.mts Generated
[log]: fixtures/sample/lib/constants.d.ts.map -> fixtures/sample/lib/constants.d.mts.map Generated
[log]: fixtures/sample/lib/constants.js -> fixtures/sample/lib/constants.mjs Generated
[log]: fixtures/sample/lib/constants.js.map -> fixtures/sample/lib/constants.mjs.map Generated
[log]: fixtures/sample/lib/database/fetch.d.ts -> fixtures/sample/lib/database/fetch.d.mts Generated
[log]: fixtures/sample/lib/database/fetch.d.ts.map -> fixtures/sample/lib/database/fetch.d.mts.map Generated
[log]: fixtures/sample/lib/database/fetch.js -> fixtures/sample/lib/database/fetch.mjs Generated
[log]: fixtures/sample/lib/database/fetch.js.map -> fixtures/sample/lib/database/fetch.mjs.map Generated
[log]: fixtures/sample/lib/database/index.d.ts -> fixtures/sample/lib/database/index.d.mts Generated
[log]: fixtures/sample/lib/database/index.d.ts.map -> fixtures/sample/lib/database/index.d.mts.map Generated
[log]: fixtures/sample/lib/database/index.js -> fixtures/sample/lib/database/index.mjs Generated
[log]: fixtures/sample/lib/database/index.js.map -> fixtures/sample/lib/database/index.mjs.map Generated
[log]: fixtures/sample/lib/index.d.ts -> fixtures/sample/lib/index.d.mts Generated
[log]: fixtures/sample/lib/index.d.ts.map -> fixtures/sample/lib/index.d.mts.map Generated
[log]: fixtures/sample/lib/index.js -> fixtures/sample/lib/index.mjs Generated
[log]: fixtures/sample/lib/index.js.map -> fixtures/sample/lib/index.mjs.map Generated
[log]: fixtures/sample/lib/lookup.d.ts -> fixtures/sample/lib/lookup.d.mts Generated
[log]: fixtures/sample/lib/lookup.d.ts.map -> fixtures/sample/lib/lookup.d.mts.map Generated
[log]: fixtures/sample/lib/lookup.js -> fixtures/sample/lib/lookup.mjs Generated
[log]: fixtures/sample/lib/lookup.js.map -> fixtures/sample/lib/lookup.mjs.map Generated
[log]: fixtures/sample/lib/types.d.ts -> fixtures/sample/lib/types.d.mts Generated
[log]: fixtures/sample/lib/types.d.ts.map -> fixtures/sample/lib/types.d.mts.map Generated
[log]: fixtures/sample/lib/types.js -> fixtures/sample/lib/types.mjs Generated
[log]: fixtures/sample/lib/types.js.map -> fixtures/sample/lib/types.mjs.map Generated
[warn]: 
[error]: 
[stdout]: 
[stderr]: 
"
`;

exports[`app > run (--dry-run) [ 'fixtures/sample/lib' ] 2`] = `
"Console Output:
[log]: done.
[log]: fixtures/sample/lib/app.d.ts -> fixtures/sample/lib/app.d.mts Generated
[log]: fixtures/sample/lib/app.d.ts.map -> fixtures/sample/lib/app.d.mts.map Generated
[log]: fixtures/sample/lib/app.js -> fixtures/sample/lib/app.mjs Generated
[log]: fixtures/sample/lib/app.js.map -> fixtures/sample/lib/app.mjs.map Generated
[log]: fixtures/sample/lib/constants.d.ts -> fixtures/sample/lib/constants.d.mts Generated
[log]: fixtures/sample/lib/constants.d.ts.map -> fixtures/sample/lib/constants.d.mts.map Generated
[log]: fixtures/sample/lib/constants.js -> fixtures/sample/lib/constants.mjs Generated
[log]: fixtures/sample/lib/constants.js.map -> fixtures/sample/lib/constants.mjs.map Generated
[log]: fixtures/sample/lib/database/fetch.d.ts -> fixtures/sample/lib/database/fetch.d.mts Generated
[log]: fixtures/sample/lib/database/fetch.d.ts.map -> fixtures/sample/lib/database/fetch.d.mts.map Generated
[log]: fixtures/sample/lib/database/fetch.js -> fixtures/sample/lib/database/fetch.mjs Generated
[log]: fixtures/sample/lib/database/fetch.js.map -> fixtures/sample/lib/database/fetch.mjs.map Generated
[log]: fixtures/sample/lib/database/index.d.ts -> fixtures/sample/lib/database/index.d.mts Generated
[log]: fixtures/sample/lib/database/index.d.ts.map -> fixtures/sample/lib/database/index.d.mts.map Generated
[log]: fixtures/sample/lib/database/index.js -> fixtures/sample/lib/database/index.mjs Generated
[log]: fixtures/sample/lib/database/index.js.map -> fixtures/sample/lib/database/index.mjs.map Generated
[log]: fixtures/sample/lib/index.d.ts -> fixtures/sample/lib/index.d.mts Generated
[log]: fixtures/sample/lib/index.d.ts.map -> fixtures/sample/lib/index.d.mts.map Generated
[log]: fixtures/sample/lib/index.js -> fixtures/sample/lib/index.mjs Generated
[log]: fixtures/sample/lib/index.js.map -> fixtures/sample/lib/index.mjs.map Generated
[log]: fixtures/sample/lib/lookup.d.ts -> fixtures/sample/lib/lookup.d.mts Generated
[log]: fixtures/sample/lib/lookup.d.ts.map -> fixtures/sample/lib/lookup.d.mts.map Generated
[log]: fixtures/sample/lib/lookup.js -> fixtures/sample/lib/lookup.mjs Generated
[log]: fixtures/sample/lib/lookup.js.map -> fixtures/sample/lib/lookup.mjs.map Generated
[log]: fixtures/sample/lib/types.d.ts -> fixtures/sample/lib/types.d.mts Generated
[log]: fixtures/sample/lib/types.d.ts.map -> fixtures/sample/lib/types.d.mts.map Generated
[log]: fixtures/sample/lib/types.js -> fixtures/sample/lib/types.mjs Generated
[log]: fixtures/sample/lib/types.js.map -> fixtures/sample/lib/types.mjs.map Generated
[warn]: 
[error]: 
[stdout]: 
[stderr]: 
"
`;

exports[`app > run (--dry-run) [ 'not_found', '--no-must-find-files' ] 1`] = `
"Console Output:
[log]: done.
[warn]: 
[error]: 
[stdout]: 
[stderr]: 
"
`;

exports[`app > run (actual) [ '.', '--root=fixtures/sample', …(1) ] 1`] = `
"Console Output:
[log]: done.
[log]: fixtures/sample/lib/app.d.ts -> temp/lib/app.d.mts Generated
[log]: fixtures/sample/lib/app.d.ts.map - copy
[log]: fixtures/sample/lib/app.d.ts.map -> temp/lib/app.d.mts.map Generated
[log]: fixtures/sample/lib/app.js -> temp/lib/app.mjs Generated
[log]: fixtures/sample/lib/app.js.map - copy
[log]: fixtures/sample/lib/app.js.map -> temp/lib/app.mjs.map Generated
[log]: fixtures/sample/lib/constants.d.ts -> temp/lib/constants.d.mts Generated
[log]: fixtures/sample/lib/constants.d.ts.map - copy
[log]: fixtures/sample/lib/constants.d.ts.map -> temp/lib/constants.d.mts.map Generated
[log]: fixtures/sample/lib/constants.js -> temp/lib/constants.mjs Generated
[log]: fixtures/sample/lib/constants.js.map - copy
[log]: fixtures/sample/lib/constants.js.map -> temp/lib/constants.mjs.map Generated
[log]: fixtures/sample/lib/database/fetch.d.ts -> temp/lib/database/fetch.d.mts Generated
[log]: fixtures/sample/lib/database/fetch.d.ts.map - copy
[log]: fixtures/sample/lib/database/fetch.d.ts.map -> temp/lib/database/fetch.d.mts.map Generated
[log]: fixtures/sample/lib/database/fetch.js -> temp/lib/database/fetch.mjs Generated
[log]: fixtures/sample/lib/database/fetch.js.map - copy
[log]: fixtures/sample/lib/database/fetch.js.map -> temp/lib/database/fetch.mjs.map Generated
[log]: fixtures/sample/lib/database/index.d.ts -> temp/lib/database/index.d.mts Generated
[log]: fixtures/sample/lib/database/index.d.ts.map - copy
[log]: fixtures/sample/lib/database/index.d.ts.map -> temp/lib/database/index.d.mts.map Generated
[log]: fixtures/sample/lib/database/index.js -> temp/lib/database/index.mjs Generated
[log]: fixtures/sample/lib/database/index.js.map - copy
[log]: fixtures/sample/lib/database/index.js.map -> temp/lib/database/index.mjs.map Generated
[log]: fixtures/sample/lib/index.d.ts -> temp/lib/index.d.mts Generated
[log]: fixtures/sample/lib/index.d.ts.map - copy
[log]: fixtures/sample/lib/index.d.ts.map -> temp/lib/index.d.mts.map Generated
[log]: fixtures/sample/lib/index.js -> temp/lib/index.mjs Generated
[log]: fixtures/sample/lib/index.js.map - copy
[log]: fixtures/sample/lib/index.js.map -> temp/lib/index.mjs.map Generated
[log]: fixtures/sample/lib/lookup.d.ts -> temp/lib/lookup.d.mts Generated
[log]: fixtures/sample/lib/lookup.d.ts.map - copy
[log]: fixtures/sample/lib/lookup.d.ts.map -> temp/lib/lookup.d.mts.map Generated
[log]: fixtures/sample/lib/lookup.js -> temp/lib/lookup.mjs Generated
[log]: fixtures/sample/lib/lookup.js.map - copy
[log]: fixtures/sample/lib/lookup.js.map -> temp/lib/lookup.mjs.map Generated
[log]: fixtures/sample/lib/types.d.ts -> temp/lib/types.d.mts Generated
[log]: fixtures/sample/lib/types.d.ts.map - copy
[log]: fixtures/sample/lib/types.d.ts.map -> temp/lib/types.d.mts.map Generated
[log]: fixtures/sample/lib/types.js -> temp/lib/types.mjs Generated
[log]: fixtures/sample/lib/types.js.map - copy
[log]: fixtures/sample/lib/types.js.map -> temp/lib/types.mjs.map Generated
[log]: fixtures/sample/out/app.d.mts -> temp/out/app.d.mts Generated
[log]: fixtures/sample/out/app.d.mts.map - copy
[log]: fixtures/sample/out/app.d.mts.map -> temp/out/app.d.mts.map Generated
[log]: fixtures/sample/out/app.d.ts.map - copy
[log]: fixtures/sample/out/app.js.map - copy
[log]: fixtures/sample/out/app.mjs -> temp/out/app.mjs Generated
[log]: fixtures/sample/out/app.mjs.map - copy
[log]: fixtures/sample/out/app.mjs.map -> temp/out/app.mjs.map Generated
[log]: fixtures/sample/out/constants.d.mts -> temp/out/constants.d.mts Generated
[log]: fixtures/sample/out/constants.d.mts.map - copy
[log]: fixtures/sample/out/constants.d.mts.map -> temp/out/constants.d.mts.map Generated
[log]: fixtures/sample/out/constants.d.ts.map - copy
[log]: fixtures/sample/out/constants.js.map - copy
[log]: fixtures/sample/out/constants.mjs -> temp/out/constants.mjs Generated
[log]: fixtures/sample/out/constants.mjs.map - copy
[log]: fixtures/sample/out/constants.mjs.map -> temp/out/constants.mjs.map Generated
[log]: fixtures/sample/out/database/fetch.d.mts -> temp/out/database/fetch.d.mts Generated
[log]: fixtures/sample/out/database/fetch.d.mts.map - copy
[log]: fixtures/sample/out/database/fetch.d.mts.map -> temp/out/database/fetch.d.mts.map Generated
[log]: fixtures/sample/out/database/fetch.d.ts.map - copy
[log]: fixtures/sample/out/database/fetch.js.map - copy
[log]: fixtures/sample/out/database/fetch.mjs -> temp/out/database/fetch.mjs Generated
[log]: fixtures/sample/out/database/fetch.mjs.map - copy
[log]: fixtures/sample/out/database/fetch.mjs.map -> temp/out/database/fetch.mjs.map Generated
[log]: fixtures/sample/out/database/index.d.mts -> temp/out/database/index.d.mts Generated
[log]: fixtures/sample/out/database/index.d.mts.map - copy
[log]: fixtures/sample/out/database/index.d.mts.map -> temp/out/database/index.d.mts.map Generated
[log]: fixtures/sample/out/database/index.d.ts.map - copy
[log]: fixtures/sample/out/database/index.js.map - copy
[log]: fixtures/sample/out/database/index.mjs -> temp/out/database/index.mjs Generated
[log]: fixtures/sample/out/database/index.mjs.map - copy
[log]: fixtures/sample/out/database/index.mjs.map -> temp/out/database/index.mjs.map Generated
[log]: fixtures/sample/out/index.d.mts -> temp/out/index.d.mts Generated
[log]: fixtures/sample/out/index.d.mts.map - copy
[log]: fixtures/sample/out/index.d.mts.map -> temp/out/index.d.mts.map Generated
[log]: fixtures/sample/out/index.d.ts.map - copy
[log]: fixtures/sample/out/index.js.map - copy
[log]: fixtures/sample/out/index.mjs -> temp/out/index.mjs Generated
[log]: fixtures/sample/out/index.mjs.map - copy
[log]: fixtures/sample/out/index.mjs.map -> temp/out/index.mjs.map Generated
[log]: fixtures/sample/out/lookup.d.mts -> temp/out/lookup.d.mts Generated
[log]: fixtures/sample/out/lookup.d.mts.map - copy
[log]: fixtures/sample/out/lookup.d.mts.map -> temp/out/lookup.d.mts.map Generated
[log]: fixtures/sample/out/lookup.d.ts.map - copy
[log]: fixtures/sample/out/lookup.js.map - copy
[log]: fixtures/sample/out/lookup.mjs -> temp/out/lookup.mjs Generated
[log]: fixtures/sample/out/lookup.mjs.map - copy
[log]: fixtures/sample/out/lookup.mjs.map -> temp/out/lookup.mjs.map Generated
[log]: fixtures/sample/out/types.d.mts -> temp/out/types.d.mts Generated
[log]: fixtures/sample/out/types.d.mts.map - copy
[log]: fixtures/sample/out/types.d.mts.map -> temp/out/types.d.mts.map Generated
[log]: fixtures/sample/out/types.d.ts.map - copy
[log]: fixtures/sample/out/types.js.map - copy
[log]: fixtures/sample/out/types.mjs -> temp/out/types.mjs Generated
[log]: fixtures/sample/out/types.mjs.map - copy
[log]: fixtures/sample/out/types.mjs.map -> temp/out/types.mjs.map Generated
[log]: fixtures/sample/src/app.ts -> temp/src/app.mts Generated
[log]: fixtures/sample/src/constants.ts -> temp/src/constants.mts Generated
[log]: fixtures/sample/src/database/fetch.ts -> temp/src/database/fetch.mts Generated
[log]: fixtures/sample/src/database/index.ts -> temp/src/database/index.mts Generated
[log]: fixtures/sample/src/index.ts -> temp/src/index.mts Generated
[log]: fixtures/sample/src/lookup.ts -> temp/src/lookup.mts Generated
[log]: fixtures/sample/src/types.ts -> temp/src/types.mts Generated
[log]: fixtures/sample/tsconfig.cjs.json - copy
[log]: fixtures/sample/tsconfig.json - copy
[warn]: 
[error]: 
[stdout]: 
[stderr]: 
"
`;

exports[`app > run error '_not_found_' 1`] = `
"Console Output:
[log]: 
[warn]: 
[error]: 
[stdout]: 
[stderr]: No files found.
[stderr]: 
"
`;

exports[`app > run error '--help' 1`] = `
"Console Output:
[log]: 
[warn]: 
[error]: 
[stdout]: Usage: ts2mjs [options] <files...>
[stdout]: 
[stdout]: Rename ESM .js files to .mjs (or .cjs if --cjs options is used)
[stdout]: 
[stdout]: Arguments:
[stdout]:   files                 Globs matching files to rename.
[stdout]: 
[stdout]: Options:
[stdout]:   --cjs                 Use .cjs extension instead of .mjs extension.
[stdout]:   -o, --output <dir>    The output directory.
[stdout]:   --cwd <dir>           The current working directory.
[stdout]:   --root <dir>          The root directory.
[stdout]:   -x,--exclude <glob>   Exclude a glob from the files. (default: [])
[stdout]:   --dry-run             Dry Run do not update files.
[stdout]:   --remove-source       Remove the original files after successful conversion.
[stdout]:   --skip-ts             Skip all TypeScript files (.ts and .d.ts) instead of
[stdout]:                         converting them.
[stdout]:   --no-must-find-files  No error if files are not found.
[stdout]:   --no-enforce-root     Do not fail if relative \`.js\` files outside of the root
[stdout]:                         are imported.
[stdout]:   --color               Force color.
[stdout]:   --no-color            Do not use color.
[stdout]:   -v, --verbose         Verbose mode
[stdout]:   -V, --version         output the version number
[stdout]:   -h, --help            display help for command
[stdout]: 
[stderr]: 
"
`;
